modules/library/main/src/main/java/org/geotools/filter/function/InterpolateFunction.java

```diff
@@ -372,29 +371,32 @@ public <T> T evaluate(Object object, Class<T> context) {
         }
     }
 
+    /**
+     * Returns the default context based on method when it's not specified (e.g. null or
+     * Object.class), for the Converters to return the default object type: Color if method ==
+     * COLOR, Double of method == NUMERIC. If a specific "context" is asked, returns it as-is.
+     */
     @SuppressWarnings("unchecked")
     private <T> Class<T> sanitizeContext(Class<T> context) {
-        if (null == context || Object.class.equals(context)) {
+        final boolean specified = context != null && !Object.class.equals(context);
+        if (!specified) {
             if (method == Method.NUMERIC) {
                 context = (Class<T>) Double.class;
             } else if (method == Method.COLOR) {
                 context = (Class<T>) Color.class;
             } else {
+                // should never reach here for as long as initialize() is called first
                 throw new IllegalStateException("Unknown method, expected NUMERIC or COLOR");
             }
         }
         return context;
     }
 
     private void validateArguments(Class<?> context) {
-        if (method == Method.NUMERIC && Color.class.isAssignableFrom(context)) {
+        if (Color.class.isAssignableFrom(context) && method != Method.COLOR) {
             throw new IllegalArgumentException(
-                    "Trying to evaluate the function as Color but the method parameter is set as NUMERIC");
-        } else if (method == Method.COLOR && !Color.class.isAssignableFrom(context)) {
-            throw new IllegalArgumentException(
-                    "Trying to evaluate the function as "
-                            + context.getSimpleName()
-                            + " but the method parameter is set as COLOR");
+                    "Trying to evaluate the function as Color but the method parameter is set as "
+                            + method);
```

Comment:
The combo of sanitizeContext (to supply a context based on method if null or object) and this validateArguments ... makes me uncomfortable but I see what you are doing and why it is working.

The error message is failing, not trying anymore:
```suggestion
                    "Unable to evaluate Color as the interpolation method is set as " + method);
```